In [1]:
from dotenv import load_dotenv
from glob import glob
from pprint import pprint #json 형식 출력에 편리
import os

In [2]:
load_dotenv()

True

## 여러 자료 형식에서 읽어오기

### pdf 문서 읽기

In [ ]:
from langchain_community.document_loaders import PyPDFLoader

pdf_loader = PyPDFLoader('./data/transformer.pdf')
pdf_docs = pdf_loader.load()

print(len(pdf_docs)) # number of documents
print()
print(type(pdf_docs)) # list of documents
print()
print(pdf_docs[0]) # document

15

<class 'list'>

page_content='Provided proper attribution is provided, Google hereby grants permission to
reproduce the tables and figures in this paper solely for use in journalistic or
scholarly works.
Attention Is All You Need
Ashish Vaswani∗
Google Brain
avaswani@google.comNoam Shazeer∗
Google Brain
noam@google.comNiki Parmar∗
Google Research
nikip@google.comJakob Uszkoreit∗
Google Research
usz@google.com
Llion Jones∗
Google Research
llion@google.comAidan N. Gomez∗ †
University of Toronto
aidan@cs.toronto.eduŁukasz Kaiser∗
Google Brain
lukaszkaiser@google.com
Illia Polosukhin∗ ‡
illia.polosukhin@gmail.com
Abstract
The dominant sequence transduction models are based on complex recurrent or
convolutional neural networks that include an encoder and a decoder. The best
performing models also connect the encoder and decoder through an attention
mechanism. We propose a new simple network architecture, the Transformer,
based solely on attention mechanisms, dispensing with recurrence a

In [ ]:
print(pdf_docs[0].page_content) # content of the document
print()
print(pdf_docs[0].metadata) # metadata of the document

Provided proper attribution is provided, Google hereby grants permission to
reproduce the tables and figures in this paper solely for use in journalistic or
scholarly works.
Attention Is All You Need
Ashish Vaswani∗
Google Brain
avaswani@google.comNoam Shazeer∗
Google Brain
noam@google.comNiki Parmar∗
Google Research
nikip@google.comJakob Uszkoreit∗
Google Research
usz@google.com
Llion Jones∗
Google Research
llion@google.comAidan N. Gomez∗ †
University of Toronto
aidan@cs.toronto.eduŁukasz Kaiser∗
Google Brain
lukaszkaiser@google.com
Illia Polosukhin∗ ‡
illia.polosukhin@gmail.com
Abstract
The dominant sequence transduction models are based on complex recurrent or
convolutional neural networks that include an encoder and a decoder. The best
performing models also connect the encoder and decoder through an attention
mechanism. We propose a new simple network architecture, the Transformer,
based solely on attention mechanisms, dispensing with recurrence and convolutions
entirely. Experime

### web 에서 읽어오기

In [10]:
from langchain_community.document_loaders import WebBaseLoader

web_loader = WebBaseLoader(['https://python.langchain.com/docs/introduction/', 'https://js.langchain.com/docs/introduction/'])

web_docs = web_loader.load()

print(len(web_docs))
print()
print(type(web_docs[0]))
print()
print(web_docs[0])

2

<class 'langchain_core.documents.base.Document'>

page_content='




Introduction | 🦜️🔗 LangChain








Skip to main contentOur new LangChain Academy Course Deep Research with LangGraph is now live! Enroll for free.IntegrationsAPI ReferenceMoreContributingPeopleError referenceLangSmithLangGraphLangChain HubLangChain JS/TSv0.3v0.3v0.2v0.1💬SearchIntroductionTutorialsBuild a Question Answering application over a Graph DatabaseTutorialsBuild a simple LLM application with chat models and prompt templatesBuild a ChatbotBuild a Retrieval Augmented Generation (RAG) App: Part 2Build an Extraction ChainBuild an AgentTaggingBuild a Retrieval Augmented Generation (RAG) App: Part 1Build a semantic search engineBuild a Question/Answering system over SQL dataSummarize TextHow-to guidesHow-to guidesHow to use tools in a chainHow to use a vectorstore as a retrieverHow to add memory to chatbotsHow to use example selectorsHow to add a semantic layer over graph databaseHow to invoke runnables in para

In [ ]:
print(web_docs[0].page_content)






Introduction | 🦜️🔗 LangChain








Skip to main contentOur new LangChain Academy Course Deep Research with LangGraph is now live! Enroll for free.IntegrationsAPI ReferenceMoreContributingPeopleError referenceLangSmithLangGraphLangChain HubLangChain JS/TSv0.3v0.3v0.2v0.1💬SearchIntroductionTutorialsBuild a Question Answering application over a Graph DatabaseTutorialsBuild a simple LLM application with chat models and prompt templatesBuild a ChatbotBuild a Retrieval Augmented Generation (RAG) App: Part 2Build an Extraction ChainBuild an AgentTaggingBuild a Retrieval Augmented Generation (RAG) App: Part 1Build a semantic search engineBuild a Question/Answering system over SQL dataSummarize TextHow-to guidesHow-to guidesHow to use tools in a chainHow to use a vectorstore as a retrieverHow to add memory to chatbotsHow to use example selectorsHow to add a semantic layer over graph databaseHow to invoke runnables in parallelHow to stream chat model responsesHow to add default invocation

In [13]:
print(web_docs[0].metadata)

{'source': 'https://python.langchain.com/docs/introduction/', 'title': 'Introduction | 🦜️🔗 LangChain', 'description': 'LangChain is a framework for developing applications powered by large language models (LLMs).', 'language': 'en'}


In [12]:
print(web_docs[1].page_content)






Introduction | 🦜️🔗 Langchain










Skip to main contentOur new LangChain Academy Course Deep Research with LangGraph is now live! Enroll for free.IntegrationsAPI ReferenceMorePeopleCommunityError referenceExternal guidesContributingv0.3v0.3v0.2v0.1🦜🔗LangSmithLangSmith DocsLangChain HubLangServePython DocsChatSearchIntroductionTutorialsBuild a Question Answering application over a Graph DatabaseTutorialsBuild a simple LLM application with chat models and prompt templatesBuild a ChatbotBuild a Retrieval Augmented Generation (RAG) App: Part 2Build an Extraction ChainTaggingBuild a Retrieval Augmented Generation (RAG) App: Part 1Build a semantic search engineBuild a Question/Answering system over SQL dataSummarize TextHow-to guidesHow-to guidesHow to add memory to chatbotsHow to use example selectorsInstallationHow to stream responses from an LLMHow to stream chat model responsesHow to embed text dataHow to use few shot examples in chat modelsHow to cache model responsesHow to cac

### csv 파일 읽어오기

In [ ]:
from langchain_community.document_loaders import CSVLoader

csv_loader = CSVLoader(file_path='./data/kbo_teams_2023.csv', encoding='utf-8')

csv_docs = csv_loader.load() #각 행을 document 로 List[Document] 형태로 반환

print("문서의 수 : ", len(csv_docs))
print("="*30)
print("첫 번째 문서의 내용 : ", csv_docs[0].page_content) # 각 행 별로 문서로 불러와 지는 것을 확인
print("="*30)
print("첫 번째 문서의 메타데이터 : ", csv_docs[0].metadata)

문서의 수 :  10
첫 번째 문서의 내용 :  Team: KIA 타이거즈
City: 광주
Founded: 1982
Home Stadium: 광주-기아 챔피언스 필드
Championships: 11
Introduction: KBO 리그의 전통 강호로, 역대 최다 우승 기록을 보유하고 있다. '타이거즈 스피릿'으로 유명하며, 양현종, 안치홍 등 스타 선수들을 배출했다. 광주를 연고로 하는 유일한 프로야구팀으로 지역 사랑이 강하다.
첫 번째 문서의 메타데이터 :  {'source': './data/kbo_teams_2023.csv', 'row': 0}


## chunk 만들기

### RecursiveCharacterTextSplitter

꼭 글자수 만이 아니라, 여러 기준으로 자를 수 있다. Separators 를 여러 개 등록할 수 있다.

In [ ]:
from langchain_text_splitters import RecursiveCharacterTextSplitter

text_splitter = RecursiveCharacterTextSplitter(
    chunk_size = 1000,
    chunk_overlap = 200, # 문맥을 살리기 위해서 겹치는 부분 크기
    length_function = len, # 글자 수를 기준 -> 이걸 왜씀? 청크 사이즈 1000이 글자수 1000이 되도록 한다.
    separators = ["\n\n", "\n", " ", ""], # 여러 기준으로 청크 나누기가 가능, 앞에 나오는 것 우선 -> 이게 RecursiveCharacterTextSplitter의 핵심 기능
)

texts = text_splitter.split_documents(pdf_docs) # 위에서 읽은 pdf
print(f"생성된 텍스트 청크 수 : {len(texts)}")
print(f"각 청크의 길이 : {list(len(text.page_content) for text in texts)}")

생성된 텍스트 청크 수 : 52
각 청크의 길이 : [981, 910, 975, 451, 932, 998, 904, 907, 995, 385, 926, 953, 216, 920, 996, 829, 975, 910, 906, 870, 929, 961, 945, 997, 195, 977, 968, 947, 933, 965, 938, 915, 733, 952, 945, 948, 618, 980, 989, 994, 624, 945, 914, 946, 918, 988, 929, 929, 849, 812, 814, 817]


In [20]:
for text in texts[:5]:
    print(text.page_content[:200])
    print("-"*40)
    print(text.page_content[-200:])
    print("="*40)

Provided proper attribution is provided, Google hereby grants permission to
reproduce the tables and figures in this paper solely for use in journalistic or
scholarly works.
Attention Is All You Need

----------------------------------------
the encoder and decoder through an attention
mechanism. We propose a new simple network architecture, the Transformer,
based solely on attention mechanisms, dispensing with recurrence and convolutions
mechanism. We propose a new simple network architecture, the Transformer,
based solely on attention mechanisms, dispensing with recurrence and convolutions
entirely. Experiments on two machine transla
----------------------------------------
the
best models from the literature. We show that the Transformer generalizes well to
other tasks by applying it successfully to English constituency parsing both with
large and limited training data.
best models from the literature. We show that the Transformer generalizes well to
other tasks by applying it succe

### 정규 표현식 분할

In [23]:
from langchain_text_splitters import CharacterTextSplitter

text_splitter = CharacterTextSplitter(
    chunk_size = 1000,
    chunk_overlap = 100,
    separator= r'(?<=[.!?])\s+', # 정규 표현식을 통한 문자열 seperator
    is_separator_regex=True,
    keep_separator = True,
)

texts = text_splitter.split_documents(pdf_docs) # 위에서 읽은 pdf
print(f"생성된 텍스트 청크 수 : {len(texts)}")
print(f"각 청크의 길이 : {list(len(text.page_content) for text in texts)}")

생성된 텍스트 청크 수 : 49
각 청크의 길이 : [991, 886, 940, 210, 907, 823, 785, 945, 880, 1000, 899, 838, 842, 865, 773, 927, 939, 527, 979, 928, 944, 594, 922, 935, 991, 534, 965, 780, 926, 524, 974, 944, 844, 291, 882, 996, 987, 400, 945, 982, 937, 617, 988, 921, 974, 548, 812, 814, 817]


In [26]:
for text in texts[:5]:
    print(text.page_content[:200])
    print("-"*40)
    print(text.page_content[-200:])
    print("="*40)

Provided proper attribution is provided, Google hereby grants permission to
reproduce the tables and figures in this paper solely for use in journalistic or
scholarly works.
Attention Is All You Need

----------------------------------------
r and decoder through an attention
mechanism. We propose a new simple network architecture, the Transformer,
based solely on attention mechanisms, dispensing with recurrence and convolutions
entirely.
Experiments on two machine translation tasks show these models to
be superior in quality while being more parallelizable and requiring significantly
less time to train. Our model achieves 28.4 BLEU on
----------------------------------------
ituency parsing both with
large and limited training data.
∗Equal contribution. Listing order is random. Jakob proposed replacing RNNs with self-attention and started
the effort to evaluate this idea.
Jakob proposed replacing RNNs with self-attention and started
the effort to evaluate this idea. Ashish, with Illia

### 허깅페이스에서 토크나이저 사용하기

In [27]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("BAAI/bge-m3")

c:\Users\USER\AppData\Local\pypoetry\Cache\virtualenvs\langchain-env-aZmMjlRy-py3.11\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\USER\.cache\huggingface\hub\models--BAAI--bge-m3. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


In [28]:
tokenizer

XLMRobertaTokenizerFast(name_or_path='BAAI/bge-m3', vocab_size=250002, model_max_length=8192, is_fast=True, padding_side='right', truncation_side='right', special_tokens={'bos_token': '<s>', 'eos_token': '</s>', 'unk_token': '<unk>', 'sep_token': '</s>', 'pad_token': '<pad>', 'cls_token': '<s>', 'mask_token': '<mask>'}, clean_up_tokenization_spaces=True),  added_tokens_decoder={
	0: AddedToken("<s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	1: AddedToken("<pad>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	2: AddedToken("</s>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	3: AddedToken("<unk>", rstrip=False, lstrip=False, single_word=False, normalized=False, special=True),
	250001: AddedToken("<mask>", rstrip=False, lstrip=True, single_word=False, normalized=False, special=True),
}

In [29]:
tokens = tokenizer.encode('안녕하세요. 반갑습니다.')

print(tokens)

[0, 107687, 5, 20451, 54272, 16367, 5, 2]


In [ ]:
print(tokenizer.convert_ids_to_tokens(tokens)) # tokens 을 가지고 id 토큰을 text 토큰으로 변환

['<s>', '▁안녕하세요', '.', '▁반', '갑', '습니다', '.', '</s>']


In [ ]:
print(tokenizer.decode(tokens)) # 다시 문자열로 디코딩

<s> 안녕하세요. 반갑습니다.</s>


## 임베딩 (Embedding)

이전까지 우리는 openAI 임베딩 모델을 사용

임베딩 : 텍스트를 벡터로 표현함으로써 의미적으로 가장 유사한 다른 텍스트를 찾는 등 수학적 연산을 수행가능하기 위해 하는 것

### openAI

In [32]:
from langchain_openai import OpenAIEmbeddings

embeddings_model = OpenAIEmbeddings(model = "text-embedding-3-small")

embeddings_model

OpenAIEmbeddings(client=<openai.resources.embeddings.Embeddings object at 0x000002846CD27A10>, async_client=<openai.resources.embeddings.AsyncEmbeddings object at 0x00000284667EAF90>, model='text-embedding-3-small', dimensions=None, deployment='text-embedding-ada-002', openai_api_version='', openai_api_base=None, openai_api_type='', openai_proxy='', embedding_ctx_length=8191, openai_api_key=SecretStr('**********'), openai_organization=None, allowed_special=None, disallowed_special=None, chunk_size=1000, max_retries=2, request_timeout=None, headers=None, tiktoken_enabled=True, tiktoken_model_name=None, show_progress_bar=False, model_kwargs={}, skip_empty=False, default_headers=None, default_query=None, retry_min_seconds=4, retry_max_seconds=20, http_client=None, http_async_client=None, check_embedding_ctx_length=True)

embedding_ctx_length=8191, 이라는 부분이 임베딩 할 수 있는 토큰의 한계, context의 한계 토큰 수

따라서 chunk 할 때도 이 이상의 토큰을 가지도록 하면 안된다.

In [34]:
embeddings_model.embedding_ctx_length

8191

In [36]:
documents = [
    "인공지능은 컴퓨터 과학의 한 분야입니다.",
    "머신러닝은 인공지능의 하위 분야입니다.",
    "딥러닝은 머신러닝의 한 종류입니다.",
    "자연어 처리는 컴퓨터가 인간의 언어를 이해하고 생성하는 기술입니다.",
    "컴퓨터 비전은 컴퓨터가 디지털 이미지나 비디오를 이해하는 방법을 연구합니다."
]

document_embeddings = embeddings_model.embed_documents(documents)

print(f"임베딩 벡터의 개수 : {len(document_embeddings)}")
print(f"임베딩 벡터의 차원 : {len(document_embeddings[0])}")
print(document_embeddings[0])

임베딩 벡터의 개수 : 5
임베딩 벡터의 차원 : 1536
[-0.002310403622686863, 0.012099752202630043, -0.002436630427837372, 0.015039036981761456, 0.018447164446115494, -0.04565808176994324, -0.003119608387351036, 0.046956416219472885, -0.018239790573716164, -0.031881313771009445, 0.006847810931503773, 0.006965021602809429, -0.018591422587633133, -0.027138788253068924, 0.004891293589025736, -0.015580009669065475, -0.041546691209077835, 0.010053073056042194, 0.05157271400094032, -0.045874472707509995, -0.01464232336729765, -0.027751890942454338, -0.018753714859485626, -0.022702815011143684, 0.004363845102488995, -0.03981557860970497, 0.05031044781208038, 0.017716851085424423, 0.00028372881934046745, -0.023333948105573654, 0.048976048827171326, -0.015273458324372768, -0.029933813959360123, -0.06159874051809311, 0.02044876292347908, 0.03537960350513458, 0.002585398033261299, -0.006807237863540649, -0.005130222998559475, 0.02477654255926609, 0.00756459916010499, 0.02753550186753273, -0.0025651114992797375, 0.025

In [37]:
embedded_query = embeddings_model.embed_query("인공지능이란 무엇인가?")

print(f"임베딩 벡터의 차원 : {len(embedded_query)}")

임베딩 벡터의 차원 : 1536


임베딩한 다큐먼트 리스트와 쿼리를 유사도 비교해서 벡터 db에서 쿼리에 연관된 내용을 찾는 과정을 직접 구현해보는 실습

In [ ]:
from langchain_community.utils.math import cosine_similarity
import numpy as np

def find_most_similar(query, documents, document_embeddings):
    query_embedding = embeddings_model.embed_query(query)
    similarities = cosine_similarity([query_embedding],document_embeddings)[0] #쿼리와 문서들간의 유사도 값 리스트
    most_similar_idx = np.argmax(similarities) # 유사도가 가장 높은 문서의 인덱스를 반환
    
    return documents[most_similar_idx], similarities[most_similar_idx]

queries = [
    "인공지능이란 무엇인가요?",
    "딥러닝과 머신러닝의 관계는 어떻게 되나요?",
    "컴퓨터가 이미지를 이해하는 방법은?"
]

for query in queries:
    most_similar_doc, similarity = find_most_similar(query, documents, document_embeddings)
    print(f"질문 : {query}")
    print(f"가장 유사한 문서 : {most_similar_doc}")
    print(f"유사도 {similarity:.4f}")
    print("="*50)

질문 : 인공지능이란 무엇인가요?
가장 유사한 문서 : 인공지능은 컴퓨터 과학의 한 분야입니다.
유사도 0.7117
질문 : 딥러닝과 머신러닝의 관계는 어떻게 되나요?
가장 유사한 문서 : 딥러닝은 머신러닝의 한 종류입니다.
유사도 0.6815
질문 : 컴퓨타가 이미지를 이해하는 방법은?
가장 유사한 문서 : 컴퓨터 비전은 컴퓨터가 디지털 이미지나 비디오를 이해하는 방법을 연구합니다.
유사도 0.5782


### Huggingface 임베딩 모델 사용하기 

이제까지 OpenAI 임베딩 모델을 사용해서 임베딩을 했는데, 단점이 요금이 나간다는 점.
Huggingface 에서 임베딩 하는 모델? 방식을 선택할 수 있다.

In [40]:
from langchain_huggingface.embeddings import HuggingFaceEmbeddings

embeddings_model = HuggingFaceEmbeddings(model_name = "BAAI/bge-m3")

embeddings_model

HuggingFaceEmbeddings(client=SentenceTransformer(
  (0): Transformer({'max_seq_length': 8192, 'do_lower_case': False}) with Transformer model: XLMRobertaModel 
  (1): Pooling({'word_embedding_dimension': 1024, 'pooling_mode_cls_token': True, 'pooling_mode_mean_tokens': False, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
), model_name='BAAI/bge-m3', cache_folder=None, model_kwargs={}, encode_kwargs={}, multi_process=False, show_progress=False)

In [41]:
document_embeddings = embeddings_model.embed_documents(documents) # 문자열 리스트를, 허깅페이스 embedding_model 을 사용, 임베딩 진행

print(f"임베딩 벡터의 개수 : {len(document_embeddings)}")
print(f"임베딩 벡터의 차원 : {len(document_embeddings[0])}")
print(document_embeddings[0])

임베딩 벡터의 개수 : 5
임베딩 벡터의 차원 : 1024
[-0.03941445052623749, 0.008764848113059998, -0.01268161553889513, 0.002453206107020378, -0.008944757282733917, -0.0073837218806147575, -0.005377359688282013, -0.009055864997208118, 0.03291522338986397, 0.0060455072671175, -0.027012888342142105, -0.027740875259041786, 0.00044411991257220507, 0.030136585235595703, 0.017242854461073875, 0.01709028333425522, 0.025524919852614403, -0.021856073290109634, -0.011341292411088943, -0.05702261999249458, -0.0003017137059941888, 0.013543128035962582, -0.00745010282844305, 0.01857445389032364, 0.0028946553356945515, 0.008630653843283653, -0.0007444894872605801, -0.028904108330607414, 0.02072780393064022, -0.02050059288740158, 0.00806991383433342, -0.026754198595881462, 0.003963088616728783, -0.01630389876663685, -0.07406219840049744, -0.033650368452072144, -0.023871466517448425, -0.034550052136182785, -0.034785956144332886, 0.005482987500727177, -0.0500335693359375, -0.0028035668656229973, -0.023146886378526688, -0.

In [42]:
embedded_query = embeddings_model.embed_query("인공지능이란 무엇인가?")

print(f"임베딩 벡터의 차원 : {len(embedded_query)}")

임베딩 벡터의 차원 : 1024


In [43]:
from langchain_community.utils.math import cosine_similarity
import numpy as np

def find_most_similar(query, documents, document_embeddings):
    query_embedding = embeddings_model.embed_query(query)
    similarities = cosine_similarity([query_embedding],document_embeddings)[0] #쿼리와 문서들간의 유사도 값 리스트
    most_similar_idx = np.argmax(similarities) # 유사도가 가장 높은 문서의 인덱스를 반환
    
    return documents[most_similar_idx], similarities[most_similar_idx]

queries = [
    "인공지능이란 무엇인가요?",
    "딥러닝과 머신러닝의 관계는 어떻게 되나요?",
    "컴퓨터가 이미지를 이해하는 방법은?"
]

for query in queries:
    most_similar_doc, similarity = find_most_similar(query, documents, document_embeddings)
    print(f"질문 : {query}")
    print(f"가장 유사한 문서 : {most_similar_doc}")
    print(f"유사도 {similarity:.4f}")
    print("="*50)

질문 : 인공지능이란 무엇인가요?
가장 유사한 문서 : 인공지능은 컴퓨터 과학의 한 분야입니다.
유사도 0.7269
질문 : 딥러닝과 머신러닝의 관계는 어떻게 되나요?
가장 유사한 문서 : 딥러닝은 머신러닝의 한 종류입니다.
유사도 0.7057
질문 : 컴퓨터가 이미지를 이해하는 방법은?
가장 유사한 문서 : 컴퓨터 비전은 컴퓨터가 디지털 이미지나 비디오를 이해하는 방법을 연구합니다.
유사도 0.6843
